src/Arcus.WebApi.Security/Authentication/CertificateAuthenticationFilter.cs

```diff
@@ -0,0 +1,171 @@
+        public CertificateAuthenticationFilter(
+            params (X509ValidationRequirement requirement, string configurationKey)[] requirements)
+        {
+            Guard.NotNull(requirements, nameof(requirements), "Sequence of requirements and their expected values should not be 'null'");
+            Guard.For<ArgumentException>(() => requirements.Any(requirement => String.IsNullOrWhiteSpace(requirement.configurationKey)), "Sequence of requirements cannot contain any configuration key that is blank");
+
+            _requirements = requirements;
+        }
+
+        /// <summary>
+        /// Called early in the filter pipeline to confirm request is authorized.
+        /// </summary>
+        /// <param name="context">The <see cref="T:Microsoft.AspNetCore.Mvc.Filters.AuthorizationFilterContext" />.</param>
+        public async Task OnAuthorizationAsync(AuthorizationFilterContext context)
+        {
+            Guard.NotNull(context, nameof(context));
+            Guard.NotNull(context.HttpContext, nameof(context.HttpContext));
+            Guard.For<ArgumentException>(() => context.HttpContext.Connection is null, "Invalid action context given without any HTTP connection");
+            Guard.For<ArgumentException>(() => context.HttpContext.RequestServices is null, "Invalid action context given without any HTTP request services");
+
+            ILogger logger = 
+                context.HttpContext.RequestServices
+                       .GetService<ILoggerFactory>()
+                       ?.CreateLogger<CertificateAuthenticationFilter>() 
+                ?? (ILogger) NullLogger.Instance;
+
+            ISecretProvider userDefinedSecretProvider = 
+                context.HttpContext.RequestServices.GetService<ICachedSecretProvider>()
+                ?? context.HttpContext.RequestServices.GetService<ISecretProvider>();
+
+            if (userDefinedSecretProvider == null)
+            {
+                throw new KeyNotFoundException(
+                    $"No configured {nameof(ICachedSecretProvider)} or {nameof(ISecretProvider)} implementation found in the request service container. "
+                    + "Please configure such an implementation (ex. in the Startup) of your application");
+            }
+
+            X509Certificate2 clientCertificate = context.HttpContext.Connection.ClientCertificate;
+            if (clientCertificate == null)
+            {
+                logger.LogWarning(
+                    "No client certificate was specified in the HTTP request while this authentication filter "
+                    + $"requires a certificate to validate on the {String.Join(", ", _requirements.Select(item => item.requirement))}");
+                
+                context.Result = new UnauthorizedResult();
+            }
+            else if (!await IsAllowedCertificate(clientCertificate, userDefinedSecretProvider, logger))
```

Comment:
Please don't inline this stuff :)

Basically, everything that is done here is what I was referring to where I would move that out of this class.

Reply:
It is strange, isn't it? How this was only a problem when the result was changed from `bool` to `Task<bool>` and an additional `await` was required.
I would love to hear more on how you find this a reason to introduce an temporary variable.

Reply:
It's actually a matter of preference and readability imho.
In most cases, I'd also prefer to create a temp variable that holds the result instead of awaiting in the if condition.  In this case however, I think I'd leave it like it is since it would require you to rewrite it as:

```csharp
else
{
     var isCertAllowed = await IsCertificateAllowedAsync(....);
     if( !isCertAllowed )
     {
```

Reply:
This has been changed and I like it a lot!